// next-universal-route/cjs/lib/NextRoute.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NextRoute = void 0;
var qs_1 = require("qs");
var generatePath_1 = require("../utils/generatePath");
var isAbsolutePath_1 = require("../utils/isAbsolutePath");
var NextRoute = (function () {
    function NextRoute(path, page, params, queryStringParams, query, options) {
        this.path = path;
        this.page = page;
        this.params = params;
        this.queryStringParams = queryStringParams;
        this.query = query;
        this.isAbsolutePath = isAbsolutePath_1.isAbsolutePath(path);
        this.options = options;
    }
    NextRoute.prototype.toAs = function () {
        if (this.isAbsolutePath) {
            return this.path;
        }
        var path = generatePath_1.generatePath(this.path, this.params);
        var queryString = qs_1.stringify(this.queryStringParams, { indices: false, encode: this.options.encode });
        return queryString ? path + "?" + queryString : path;
    };
    NextRoute.prototype.toHref = function () {
        if (this.isAbsolutePath) {
            return this.path;
        }
        var queryString = qs_1.stringify(__assign(__assign(__assign({}, this.query), this.params), this.queryStringParams), { indices: false, encode: this.options.encode });
        return queryString ? this.page + "?" + queryString : this.page;
    };
    return NextRoute;
}());
exports.NextRoute = NextRoute;
