// next-universal-route/es/lib/NextRoute.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { stringify } from 'qs';
import { generatePath } from '../utils/generatePath';
import { isAbsolutePath } from '../utils/isAbsolutePath';
var NextRoute = (function () {
    function NextRoute(path, page, params, queryStringParams, query, options) {
        this.path = path;
        this.page = page;
        this.params = params;
        this.queryStringParams = queryStringParams;
        this.query = query;
        this.isAbsolutePath = isAbsolutePath(path);
        this.options = options;
    }
    NextRoute.prototype.toAs = function () {
        if (this.isAbsolutePath) {
            return this.path;
        }
        var path = generatePath(this.path, this.params);
        var queryString = stringify(this.queryStringParams, { indices: false, encode: this.options.encode });
        return queryString ? path + "?" + queryString : path;
    };
    NextRoute.prototype.toHref = function () {
        if (this.isAbsolutePath) {
            return this.path;
        }
        var queryString = stringify(__assign(__assign(__assign({}, this.query), this.params), this.queryStringParams), { indices: false, encode: this.options.encode });
        return queryString ? this.page + "?" + queryString : this.page;
    };
    return NextRoute;
}());
export { NextRoute };
